sql: support BYPASSRLS role option for RLS policies

spilchen opened the pull request:
Previously, attempts to use the BYPASSRLS role option with row-level security (RLS) policies resulted in an unimplemented error. This change lifts those restrictions and adds proper support for roles or system privileges with the BYPASSRLS flag, allowing them to bypass RLS policies on tables where it is enabled.

Closes #136910

Epic: CRDB-11724
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143885)
<!-- Reviewable:end -->

spilchen committed:
sql: support BYPASSRLS role option for RLS policies

Previously, attempts to use the BYPASSRLS role option with row-level
security (RLS) policies resulted in an unimplemented error. This change
lifts those restrictions and adds proper support for roles or system
privileges with the BYPASSRLS flag, allowing them to bypass RLS policies
on tables where it is enabled.

Closes #136910

Epic: CRDB-11724
Release note: none

rafiss reviewed (approved):
lgtm! and thanks for the exhaustive test cases

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143885#-:-ONGytN18RXuCeM6oST3:b-38urat)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)



<!-- Sent from Reviewable.io -->
